src/primitives.es6

```diff
@@ -0,0 +1,339 @@
+
+    return this;
+  }
+
+  fromCompactRefract(tuple) {
+    this.element = tuple[0];
+    this.meta = tuple[1];
+    this.attributes = tuple[2];
+    this.content = (tuple[3] || []).map((content) =>
+      convertFromCompactRefract(content));
+
+    this.convertAttributesToElements(convertFromCompactRefract);
+
+    return this;
+  }
+
+  get(index) {
+    return index === undefined ? this : this.content[index];
+  }
+
+  set(index, value) {
+    this.content[index] = convertToType(value);
+    return this;
+  }
+
+  map(cb) {
+    return this.content.map(cb);
+  }
+
+  filter(condition) {
+    const newArray = new Collection();
+    newArray.content = this.content.filter(condition);
+    return newArray;
+  }
+
+  forEach(cb) {
+    this.content.forEach(cb);
+  }
+
+  push(value) {
+    this.content.push(convertToType(value));
+    return this;
+  }
+
+  add(value) {
+    this.push(value);
+  }
+
+  findElements(condition, options={}) {
+    const recursive = options.recursive === undefined ? false : options.recursive;
+    const results = options.results === undefined ? [] : options.results;
+
+    this.content.forEach((el) => {
+      // We use duck-typing here to support any registered class that
+      // may contain other elements.
+      if (recursive && (el.findElements !== undefined)) {
+          el.findElements(condition, {results, recursive});
+      }
+      if (condition(el)) {
+        results.push(el);
+      }
+    });
+    return results;
+  }
+
+  find(condition, options) {
+    const newArray = new Collection();
+    newArray.content = this.findElements(condition, options);
+    return newArray;
+  }
+}
+
+export class ArrayType extends Collection {
+  constructor(content=[], meta={}, attributes={}) {
+    const converted = content.map((value) => convertToType(value));
+
+    super('array', converted, meta, attributes);
+  }
+}
+
+export class ObjectType extends Collection {
+  constructor(content={}, meta={}, attributes={}) {
+    const converted = Object.keys(content).map((key) => {
+      const element = convertToType(content[key]);
+      element.meta.name = key;
+      return element;
+    });
+    super('object', converted, meta, attributes);
+  }
+
+  toValue() {
+    return this.content.reduce((results, el) => {
+      results[el.meta.name] = el.toValue();
+      return results;
+    }, {});
+  }
+
+  get(name) {
+    return name === undefined ? this : _.first(
+      this.content.filter((value) => value.meta.name === name)
+    );
+  }
+
+  set(name, value) {
+    const location = this.content.map(i => i.meta.name).indexOf(name);
+
+    value = convertToType(value);
+    // TODO: Should we mutate or copy here? Of course it doesn't matter
+    //       for non-refracted elements as they get copied anyway, but
+    //       if the input is already refracted and we add it to multiple
+    //       objects with a different name suddenly we have a problem.
+    //       We have the same problem in the constructor above.
```

Comment:
Good point. I'm wondering if we can handle this someway internally with a special kind of element wrapper. It would simply wrap an element in a property name.

Just throwing out ideas. Would be nice to have our cake and eat it, too.